vim-patch:8.1.{27,32,36,69,70,71,91,92}

erw7 opened the pull request:
**vim-patch:8.1.0027: difficult to make a plugin that feeds a line to a job**
Problem:    Difficult to make a plugin that feeds a line to a job.
Solution:   Add the nitial code for the "prompt" buftype.
https://github.com/vim/vim/commit/f273245f6433d5d43a5671306b520a3230c35787

**vim-patch:8.1.0032: BS in prompt buffer starts new line**
Problem:    BS in prompt buffer starts new line.
Solution:   Do not allows BS over the prompt.  Make term_sendkeys() handle special keys. Add a test.
https://github.com/vim/vim/commit/6b810d92a9cd9378ab46ea0db07079cb789f9faa

~**vim-patch:8.1.0035: not easy to switch between prompt buffer and other windows**
Problem:    Not easy to switch between prompt buffer and other windows.
Solution:   Accept CTRL-W commands in Insert mode.  Start and stop Insert mode as one would expect.
https://github.com/vim/vim/commit/6d41c78e353b630bc1a72cbff9160311d2a81e8c~

**vim-patch:8.1.0036: not restoring Insert mode if leaving prompt buffer with mouse**
Problem:    Not restoring Insert mode if leaving a prompt buffer by using a mouse click.
Solution:   Set b_prompt_insert appropriately. Also correct cursor position when moving cursor to last line.
https://github.com/vim/vim/commit/891e1fd894720d0b99a9daefa41e8181844f819a

~**vim-patch:8.1.0042: if omni completion opens a window Insert mode is stopped**
Problem:    If omni completion opens a window Insert mode is stopped. (Hirohito Higashi)
Solution:   Only set stop_insert_mode in a prompt buffer window.
https://github.com/vim/vim/commit/f98b845dd185dfadfa7a622a42452bfa6809d4e0~

~**vim-patch:8.1.0064: typing CTRL-W in a prompt buffer shows mode "-- --"**
Problem:    Typing CTRL-W in a prompt buffer shows mode "-- --".
Solution:   Set restart_edit to 'A' and check for it.
https://github.com/vim/vim/commit/942b4541a2d8e8df8369ab70e112dbbbe0c7c0aa~

**vim-patch:8.1.0069: cannot handle pressing CTRL-C in a prompt buffer**
Problem:    Cannot handle pressing CTRL-C in a prompt buffer.
Solution:   Add prompt_setinterrupt().
https://github.com/vim/vim/commit/0e5979a6d491f68c4a8c86fab489016919329a6b

**vim-patch:8.1.0070: missing part of the changes for prompt_setinterrupt()**
Problem:    Missing part of the changes for prompt_setinterrupt().
Solution:   Add the missing changes.
https://github.com/vim/vim/commit/222cd20e2662e7478cfe42b78cc4f1c153ca819d

**vim-patch:8.1.0071: terminal debugger only works with the terminal feature**
Problem:    Terminal debugger only works with the terminal feature.
Solution:   Make it also work with a prompt buffer.  Makes it possible to use on MS-Windows. Various other improvements. (closes vim/vim#3012)
https://github.com/vim/vim/commit/b3307b5e7e7bd3962b0d5c61a94e638564c146b0

**vim-patch:8.1.0091: MS-Windows: Cannot interrupt gdb when program is running**
Problem:    MS-Windows: Cannot interrupt gdb when program is running.
Solution:   Add debugbreak() and use it in the terminal debugger.  Respect 'modified' in a prompt buffer.
https://github.com/vim/vim/commit/4551c0a9fcdbdef52836d4852686d54b5e47fdaf

**vim-patch:8.1.0092: prompt buffer test fails**
Problem:    Prompt buffer test fails.
Solution:   Set 'nomodified' before closing the window. (Ozaki Kiichi, closes vim/vim#3051)
https://github.com/vim/vim/commit/71ef1ba5e996f34d3e0acbe1d89c4c6bfa5e98ba

erw7 committed:
vim-patch:8.1.0027: difficult to make a plugin that feeds a line to a job

Problem:    Difficult to make a plugin that feeds a line to a job.
Solution:   Add the nitial code for the "prompt" buftype.
https://github.com/vim/vim/commit/f273245f6433d5d43a5671306b520a3230c35787

erw7 committed:
vim-patch:8.1.0032: BS in prompt buffer starts new line

Problem:    BS in prompt buffer starts new line.
Solution:   Do not allows BS over the prompt.  Make term_sendkeys() handle
            special keys. Add a test.
https://github.com/vim/vim/commit/6b810d92a9cd9378ab46ea0db07079cb789f9faa

erw7 committed:
vim-patch:8.1.0036: not restoring Insert mode if leaving prompt buffer with mouse

Problem:    Not restoring Insert mode if leaving a prompt buffer by using a
            mouse click.
Solution:   Set b_prompt_insert appropriately. Also correct cursor position
            when moving cursor to last line.
https://github.com/vim/vim/commit/891e1fd894720d0b99a9daefa41e8181844f819a

erw7 committed:
vim-patch:8.1.0069: cannot handle pressing CTRL-C in a prompt buffer

Problem:    Cannot handle pressing CTRL-C in a prompt buffer.
Solution:   Add prompt_setinterrupt().
https://github.com/vim/vim/commit/0e5979a6d491f68c4a8c86fab489016919329a6b

erw7 committed:
vim-patch:8.1.0070: missing part of the changes for prompt_setinterrupt()

Problem:    Missing part of the changes for prompt_setinterrupt().
Solution:   Add the missing changes.
https://github.com/vim/vim/commit/222cd20e2662e7478cfe42b78cc4f1c153ca819d

erw7 committed:
vim-patch:8.1.0071: terminal debugger only works with the terminal feature

Problem:    Terminal debugger only works with the terminal feature.
Solution:   Make it also work with a prompt buffer.  Makes it possible to use
            on MS-Windows. Various other improvements. (closes vim/vim#3012)
https://github.com/vim/vim/commit/b3307b5e7e7bd3962b0d5c61a94e638564c146b0

erw7 committed:
vim-patch:8.1.0091: MS-Windows: Cannot interrupt gdb when program is running

Problem:    MS-Windows: Cannot interrupt gdb when program is running.
Solution:   Add debugbreak() and use it in the terminal debugger.
            Respect 'modified' in a prompt buffer.
https://github.com/vim/vim/commit/4551c0a9fcdbdef52836d4852686d54b5e47fdaf

erw7 committed:
vim-patch:8.1.0092: prompt buffer test fails

Problem:    Prompt buffer test fails.
Solution:   Set 'nomodified' before closing the window. (Ozaki Kiichi,
            closes vim/vim#3051
https://github.com/vim/vim/commit/71ef1ba5e996f34d3e0acbe1d89c4c6bfa5e98ba

bfredl commented on src/nvim/channel.c:
```diff
@@ -834,3 +836,55 @@ Array channel_all_info(void)
+    curwin->w_cursor.lnum = lnum + 1;
+    curwin->w_cursor.col = 0;
+
+    if (curbuf->b_prompt_callback.type == kCallbackNone
```
Checking  `curbuf->b_prompt_callback.type == kCallbackNone` should be enough.  There is no point to set `callback.type = kCallbackFuncref` without an actual funcref.

erw7 replied:
@bfredl No, this check is necessary because you can call something like `:call prompt_setcallback(bufnr(''), '')`. If we check only `curbuf->b_prompt_callback.type=kCallbackNone`, `E117: Unknown function: ` occurs.

bfredl replied:
>  However, in the case of :call prompt_setcallback(bufnr(''),''), vim performs the same check so that E117 does not occur. 

Which simply implies that `prompt_setcallback(bufnr(''),'')` should set `kCallbackNone`. The entire point of having callback being an enum is that the enum label should indicate what state the callback actually is in. If both kCallbackNone and kCallbackFuncref can be used used to indicate the valid state of "no callback", both labels become quite meaningless (and sounds like a recipe for future bugs).

> I think that is good if it is a story of changing callback_from_typval() not to allow empty string. But shouldn't it be another PR if we make that change?

This is unrelated to what I suggested. `.funcref=""`, while unfortunate, is never confused for a valid "no callback" in any existing code, what I know of.

bfredl commented on src/nvim/channel.c:
```diff
@@ -834,3 +836,55 @@ Array channel_all_info(void)
   return ret;
 }
+
+void invoke_prompt_callback(void)
```
This doesn't seem to have anything to do with (nvim) channels. Perhaps it should be in `eval.c`

jpfeiffer16 commented:
Hi! Is there a planned date for this being merged? Sorry if I missed any docs. I didn't see anything on it.

justinmk commented on runtime/doc/nvim_terminal_emulator.txt:
```diff
@@ -307,6 +307,23 @@ Other commands ~
+						*termdebug-prompt*
+When on MS-Windows, gdb will run in a buffer with 'buftype' set to "prompt".
+This works slightly differently:
+- The gdb window will be in Insert mode while typing commands.  Go to Normal
```
why does it need to be different on Windows?

erw7 replied:
Communicating with `gdb` requires a `pty`, but Windows can not get a `pty`. Therefore, Windows needs prompt buffer.

justinmk replied:
I thought that winpty could be used for that.

erw7 replied:
> I thought that winpty could be used for that.

If `winpty` provides a named pipe connected to a hidden console to pass to `gdb`, there is possibility to work. However, the current `winpty` does not provide such a pipe. Therefore, I think that does not work with the current `winpty`.

justinmk commented on src/nvim/edit.c:
```diff
@@ -950,6 +973,15 @@ static int insert_handle_key(InsertState *s)
+    if (bt_prompt(curbuf) && (mod_mask & MOD_MASK_SHIFT) == 0) {
+      // In a prompt window CTRL-W is used for window commands.
+      // Use Shift-CTRL-W to delete a word.
+      stuffcharReadbuff(Ctrl_W);
```
Nvim doesn't do this for `:terminal` buffers, it seems rather inconsistent to now do it for prompt buffers.

erw7 replied:
For now, I have removed the commit related to the move by `Ctrl_W` in insert mode from this PR.

justinmk commented on src/nvim/eval.c:
```diff
@@ -8043,6 +8049,30 @@ static void f_cursor(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+      rettv->vval.v_number = OK;
+    }
+#else
+    uv_kill(pid, SIGINT);
```
are you sure this doesn't work on Windows?

erw7 replied:
Yes, `uv_kill(pid, SIGINT);` can not be used on Windows since `gdb` is terminated instead of the program being debugged.

justinmk replied:
I was hoping that libuv translated SIGINT for us :)

erw7 commented:
In this PR, vim-8.1.0036 is only partially applied. If we import vim-8.1.0035 in the future, we will also need to import the rest of vim-8.1.0036.

justinmk commented:
Scheduled for 0.5. @erw7 please leave a comment if this sits around for too long.

erw7 committed:
Fix issue where callbacks are garbage collected

Import necessary part of vim-patch:8.1.1575.
https://github.com/vim/vim/commit/75a1a9415b9c207de5a29b25c0d1949c6c9c5c61

doronbehar commented:
Hey,

Just wanted to point out that a very popular and useful Vim only plugin [vimspector](https://github.com/puremourning/vimspector) can't work in Neovim because of these patches not merged. [A PR](https://github.com/puremourning/vimspector/pull/30) is waiting.

blueyed commented:
@doronbehar 
Have you tried if it works with this branch/PR then?

doronbehar commented:
On NixOS, the build failed with:

<details>

```
[100%] Building C object src/nvim/CMakeFiles/nvim.dir/terminal.c.o
/build/source/src/nvim/terminal.c: In function 'terminal_init':
/build/source/src/nvim/terminal.c:178:12: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'red'
       color.red = 0;
            ^
/build/source/src/nvim/terminal.c:179:12: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'green'
       color.green = 0;
            ^
/build/source/src/nvim/terminal.c:180:12: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'blue'
       color.blue = (uint8_t)(color_index + 1);
            ^
In file included from /build/source/src/nvim/func_attr.h:44,
                 from /build/source/build/include/log.h.generated.h:4,
                 from /build/source/src/nvim/log.h:72,
                 from /build/source/src/nvim/terminal.c:46:
/build/source/src/nvim/terminal.c:185:33: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'red'
                       RGB_(color.red, color.green, color.blue),
                                 ^
/build/source/src/nvim/macros.h:154:25: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                         ^
/build/source/src/nvim/terminal.c:185:44: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'green'
                       RGB_(color.red, color.green, color.blue),
                                            ^
/build/source/src/nvim/macros.h:154:37: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                                     ^
/build/source/src/nvim/terminal.c:185:57: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'blue'
                       RGB_(color.red, color.green, color.blue),
                                                         ^
/build/source/src/nvim/macros.h:154:47: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                                               ^
/build/source/src/nvim/terminal.c:191:26: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'red'
   default_vt_fg = RGB_(fg.red, fg.green, fg.blue);
                          ^
/build/source/src/nvim/macros.h:154:25: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                         ^
/build/source/src/nvim/terminal.c:191:34: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'green'
   default_vt_fg = RGB_(fg.red, fg.green, fg.blue);
                                  ^
/build/source/src/nvim/macros.h:154:37: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                                     ^
/build/source/src/nvim/terminal.c:191:44: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'blue'
   default_vt_fg = RGB_(fg.red, fg.green, fg.blue);
                                            ^
/build/source/src/nvim/macros.h:154:47: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                                               ^
/build/source/src/nvim/terminal.c:192:26: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'red'
   default_vt_bg = RGB_(bg.red, bg.green, bg.blue);
                          ^
/build/source/src/nvim/macros.h:154:25: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                         ^
/build/source/src/nvim/terminal.c:192:34: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'green'
   default_vt_bg = RGB_(bg.red, bg.green, bg.blue);
                                  ^
/build/source/src/nvim/macros.h:154:37: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                                     ^
/build/source/src/nvim/terminal.c:192:44: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'blue'
   default_vt_bg = RGB_(bg.red, bg.green, bg.blue);
                                            ^
/build/source/src/nvim/macros.h:154:47: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                                               ^
/build/source/src/nvim/terminal.c: In function 'terminal_open':
/build/source/src/nvim/terminal.c:263:12: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'red'
       color.red = 0;
            ^
/build/source/src/nvim/terminal.c:264:12: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'green'
       color.green = 0;
            ^
/build/source/src/nvim/terminal.c:265:12: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'blue'
       color.blue = (uint8_t)(i + 1);
            ^
/build/source/src/nvim/terminal.c:289:14: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'red'
         color.red = (uint8_t)((color_val >> 16) & 0xFF);
              ^
/build/source/src/nvim/terminal.c:290:14: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'green'
         color.green = (uint8_t)((color_val >> 8) & 0xFF);
              ^
/build/source/src/nvim/terminal.c:291:14: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'blue'
         color.blue = (uint8_t)((color_val >> 0) & 0xFF);
              ^
In file included from /build/source/src/nvim/func_attr.h:44,
                 from /build/source/build/include/log.h.generated.h:4,
                 from /build/source/src/nvim/log.h:72,
                 from /build/source/src/nvim/terminal.c:46:
/build/source/src/nvim/terminal.c: In function 'terminal_get_line_attributes':
/build/source/src/nvim/terminal.c:633:29: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'red'
     int vt_fg = RGB_(cell.fg.red, cell.fg.green, cell.fg.blue);
                             ^
/build/source/src/nvim/macros.h:154:25: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                         ^
/build/source/src/nvim/terminal.c:633:42: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'green'
     int vt_fg = RGB_(cell.fg.red, cell.fg.green, cell.fg.blue);
                                          ^
/build/source/src/nvim/macros.h:154:37: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                                     ^
/build/source/src/nvim/terminal.c:633:57: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'blue'
     int vt_fg = RGB_(cell.fg.red, cell.fg.green, cell.fg.blue);
                                                         ^
/build/source/src/nvim/macros.h:154:47: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                                               ^
/build/source/src/nvim/terminal.c:634:29: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'red'
     int vt_bg = RGB_(cell.bg.red, cell.bg.green, cell.bg.blue);
                             ^
/build/source/src/nvim/macros.h:154:25: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                         ^
/build/source/src/nvim/terminal.c:634:42: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'green'
     int vt_bg = RGB_(cell.bg.red, cell.bg.green, cell.bg.blue);
                                          ^
/build/source/src/nvim/macros.h:154:37: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                                     ^
/build/source/src/nvim/terminal.c:634:57: error: 'VTermColor' {aka 'union <anonymous>'} has no member named 'blue'
     int vt_bg = RGB_(cell.bg.red, cell.bg.green, cell.bg.blue);
                                                         ^
/build/source/src/nvim/macros.h:154:47: note: in definition of macro 'RGB_'
 #define RGB_(r, g, b) ((r << 16) | (g << 8) | b)
                                               ^
make[2]: *** [src/nvim/CMakeFiles/nvim.dir/build.make:2741: src/nvim/CMakeFiles/nvim.dir/terminal.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:3314: src/nvim/CMakeFiles/nvim.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
```
</details>

blueyed commented:
@doronbehar 
> On NixOS, the build failed with:

Not related to this PR/patch, is it?
Looks like a mismatch with the libvterm version.. 0.1 is required with Neovim 0.4 - CMake sure warn you about it.  Please create a new issue / provide more info if that's not the case with how you're building / and/or if this can be improved then.

blueyed commented:
@erw7 
Rebased this on master (testing cannot hurt since it is old already), and it helps with https://github.com/neovim/neovim/issues/11099#issuecomment-535235395.  /cc @doronbehar

doronbehar commented:
> @doronbehar
Have you tried if it works with this branch/PR then?

@blueyed I'm afraid not, [that vimspector PR](https://github.com/puremourning/vimspector/pull/30) is not ready yet but I've notified it's author about my attempt to try it along with _this_ PR merged.

I hope that never the less the maintainers will consider merging this sooner then in `0.5.0`.

eyalk5 commented:
Just a friendly reminder

tbodt commented:
Ping, this seems to be the main blocker for Neovim support in Vimspector

Billy4195 commented:
@erw7 @blueyed 
Because this PR has stuck for a while, I have already solved the conflicting files, and push to my own [repository@pr10433](https://github.com/Billy4195/neovim/tree/pr10433). Would you like to use my branch to update your branch to make this PR be merged or I should create a new PR.

erw7 commented:
@Billy4195 Thanks. I have updated this branch with your branch.

justinmk commented:
Merging this since it includes various useful changes (e.g. the `op_is_change` semantics is useful, and Termdebug changes), though I am not in favor of some of the surface changes:

- Introducing another special-case buffer (`buftype=prompt`)  is questionable. Ideally, plugin authors would instead have better control over normal buffers.
- `debugbreak()` is over-specific, why not something like `chansignal()` instead?

Thanks @erw7 and @Billy4195
